io/pci/pci_hotplug.py

```diff
@@ -47,76 +47,82 @@ def setUp(self):
         if cpu._list_matches(open('/proc/cpuinfo').readlines(),
                              'platform\t: pSeries\n'):
-            power_vm = True
+            self.power_vm = True
             for mdl in ['rpaphp', 'rpadlpar_io']:
                 if not linux_modules.module_is_loaded(mdl):
                     linux_modules.load_module(mdl)
         elif cpu._list_matches(open('/proc/cpuinfo').readlines(),
                                'platform\t: PowerNV\n'):
-            power_vm = False
+            self.power_vm = False
             if not linux_modules.module_is_loaded("pnv_php"):
                 linux_modules.load_module("pnv_php")
         self.return_code = 0
         self.device = self.params.get('pci_device', default=' ')
-        self.num_of_hotplug = int(self.params.get('num_of_hotplug', default='1'))
-        if not os.path.isdir('/sys/bus/pci/devices/%s' % self.device):
-            self.cancel("PCI device given does not exist")
-        self.num_of_hotplug = int(self.params.get('num_of_hotplug', default='1'))
-        if power_vm:
+        self.count = int(self.params.get('count', default='1'))
+
+    def get_slot(self, pci_addr):
+        '''
+        Returns the slot number with pci_address
+        '''
+        if self.power_vm:
             devspec = genio.read_file("/sys/bus/pci/devices/%s/devspec"
-                                      % self.device)
-            self.slot = genio.read_file("/proc/device-tree/%s/ibm,loc-code"
-                                        % devspec)
-            self.slot = re.match(r'((\w+)[\.])+(\w+)-P(\d+)-C(\d+)|Slot(\d+)',
-                                 self.slot).group()
+                                      % pci_addr)
+            slot = genio.read_file("/proc/device-tree/%s/ibm,loc-code"
+                                   % devspec)
+            slot = re.match(r'((\w+)[\.])+(\w+)-P(\d+)-C(\d+)|Slot(\d+)',
+                            slot).group()
         else:
-            self.slot = pci.get_pci_prop(self.device, "PhySlot")
-        if not os.path.isdir('/sys/bus/pci/slots/%s' % self.slot):
-            self.cancel("%s Slot not available" % self.slot)
-        if not os.path.exists('/sys/bus/pci/slots/%s/power' % self.slot):
-            self.cancel("%s Slot does not support hotplug" % self.slot)
+            slot = pci.get_pci_prop(pci_addr, "PhySlot")
+        if not os.path.isdir('/sys/bus/pci/slots/%s' % slot):
+            self.cancel("%s Slot not available" % slot)
+        if not os.path.exists('/sys/bus/pci/slots/%s/power' % slot):
+            self.cancel("%s Slot does not support hotplug" % slot)
+        return slot
 
     def test(self):
         """
         Creates namespace on the device.
         """
+        for pci_addr in self.device.split(","):
+            if not os.path.isdir('/sys/bus/pci/devices/%s' % pci_addr):
+                self.cancel(" %s does not exist" % pci_addr)
+            slot = self.get_slot(pci_addr)
```

Comment:
This can go into `setUp()`

Reply:
if we move this to setup then we need to store the slot values in dic and then run the for loop over the slot. because  we need the slot and pci_addr in other functions too like remove and add.
let me know your thought or way of handling.

Reply:
How about this then:
```
        for pci_addr in self.device.split(","):
            if not os.path.isdir('/sys/bus/pci/devices/%s' % pci_addr):
                continue
            slot = self.get_slot(pci_addr)
```

Reply:
any how, we are going to store the failed pci_add in a list and check at the last.
so I feel instead of continue I can used "err_list.append(pci_addr)" will help globally.

Reply:
yes, that is better. But if you want to `cancel`, it is better if it is `setUp()`.

Reply:
we will remove the cancel from test and will add the following line in setup for initial check and cancel

if not self.device:
    self.cancel("PCI_addr not given")

how is this?

Reply:
yes

Reply:
yes